### PR TITLE
feat(pie-modal): Modal heading props

### PR DESCRIPTION
**Pull request by raoufswe**
## Describe your changes (can list changeset entries if preferable)
- adds heading props to modal
- modifies the react-wrapper script to take care of cases where the custom element is defined via a variable

## Author Checklist (complete before requesting a review)
- [x] I have performed a self-review of my code
- [x] If it is a core feature, I have added thorough tests
- [ ] If it is a `PIE Docs` change, I have reviewed the Docs site preview
- [x] If it is a component change, I have reviewed the Storybook preview
- [x] If there are visual test updates, I have reviewed them properly before approving

### Reviewer 1
- [ ] If it is a `PIE Docs` change, I have reviewed the PR preview
- [ ] If there are visual test updates, I have reviewed them

### Reviewer 2
- [ ] If it is a `PIE Docs` change, I have reviewed the PR preview
- [ ] If there are visual test updates, I have reviewed them

### CODE REVIEW

File: packages/components/pie-modal/README.md
Comment (kevinrodrigues) [inline packages/components/pie-modal/README.md:55]: I think we should add in another column to state which props are required. Then you can mark the props you added as required `yes`

Comment (raoufswe) [reply packages/components/pie-modal/README.md:55]: I thought about the same thing but then decided to use `*` to indicate a prop is required as there will be few required props compared to optional and the table might be a bit hard to digest. Skip's [storybook](https://react.pie.design/?path=/story/components-modal--playground) has a similar table structure. 

as a side note, ideally, we should generate the metadata through jsDoc or the typescript definition as this approach will get out of sync easily but something to keep in mind

**Review by kevinrodrigues — DISMISSED** [review]

File: packages/components/pie-modal/README.md
Comment (jamieomaguire) [reply packages/components/pie-modal/README.md:55]: Fully agree - we should automate this!

File: packages/components/pie-modal/src/index.ts
Comment (jamieomaguire) [inline packages/components/pie-modal/src/index.ts:20]: if it's required then I don't think it can be `undefined`, right?

File: packages/components/pie-modal/src/index.ts
Comment (jamieomaguire) [inline packages/components/pie-modal/src/index.ts:35]: could you potentially do:

```
<h${headingLevel}>${heading}</h${headingLevel>
```

This way you wouldn't need to call `unsafeStatic`?

Comment (jamieomaguire) [reply packages/components/pie-modal/src/index.ts:35]: in this example, heading level would be a string number like '1', '2' etc

File: packages/components/pie-modal/test/visual/pie-modal.spec.ts
Comment (jamieomaguire) [inline packages/components/pie-modal/test/visual/pie-modal.spec.ts:6]: whilst this is perfectly reasonable, if you need to label the tests I would suggest using the test component wrapper and providing a string label.

Entirely up to you but it might be nicer for consistency. Example can be found in the pie-button visual tests

File: packages/components/pie-modal/src/index.ts
Comment (raoufswe) [reply packages/components/pie-modal/src/index.ts:20]: I agree, but I get this error if I don't specify if the component potentially won't be assigned, do you have any idea?
Screenshot 2023-06-09 at 16 43 03

**Comment by jamieomaguire** [comment]
> Screenshot 2023-06-09 at 15 47 00
> Storybook doesn't appear to be working @raoufswe

@raoufswe it actually seems like a refresh fixes this but it broke for me when I clicked on each component and had to refresh. Does this occur for you too?

**Comment by jamieomaguire** [comment]
> @jamieomaguire, yes, it's the issue I shared with you yesterday. the core package is imported several times and it declares the test custom element every time it's imported, thus, the error.

yes that makes sense - can we add a temporary registry check then pull that test code into a separate package next week as discussed?

File: packages/components/pie-modal/src/index.ts
Comment (jamieomaguire) [reply packages/components/pie-modal/src/index.ts:20]: I don't have an immediate suggestion but maybe we can look into this a bit more as I think it could make the codebase messy if we set a precedent here

File: packages/components/pie-modal/src/index.ts
Comment (raoufswe) [reply packages/components/pie-modal/src/index.ts:35]: Unfortunately, I believe you still need to use `unsafeStatic` as it's still considered a static expression. I find [the lit doc ](https://lit.dev/docs/templates/expressions/#static-expressions)about the topic quite useful. but in a nutshell, we have a validation layer before using `unsafeStatic ` and only valid heading tags will be passed through with the help of `@validPropertyValues` 

Screenshot 2023-06-09 at 16 50 57

Comment (jamieomaguire) [reply packages/components/pie-modal/src/index.ts:37]: ah, that makes sense. I'm still unsure if I like the idea of a vanilla consumer having to write:

`<pie-modal heading-level="h2" />`

but it's probably fine!

Comment (raoufswe) [reply packages/components/pie-modal/src/index.ts:37]: It's very common to pass the html tag as a prop as far as I know, tho I would argue the benefits such flexibility will provide to our users, but I assume it was in the requirements for a reason. 

I'm happy to make the prop numerical if it makes more sense and the rest of the team 🚀💯

File: packages/components/pie-modal/src/defs.ts
Comment (raoufswe) [reply packages/components/pie-modal/src/defs.ts:4]: I will check before merging the PR if the union literal work is ready or not, if not, I reckon we shouldn't block the PR because of that and assume a clean up will needed 👍

File: packages/components/pie-modal/src/index.ts
Comment (raoufswe) [reply packages/components/pie-modal/src/index.ts:20]: I did a bit of digging and I believe we can use something called _definite assignment assertion operator_ or `!` in short to achieve a semi-required behavior in classes and ts. 

assuming the title is a required field and should always be string, setting `heading!: string` will tell the compiler that the heading property will be instantiated with a value before it's used, even though it might not be apparent for you now, but every time we get or set that property, we want it to be treated it as a type of string. 👓

File: packages/components/pie-modal/test/visual/pie-modal.spec.ts
Comment (raoufswe) [reply packages/components/pie-modal/test/visual/pie-modal.spec.ts]: actually, I might've just been a bit expressive here and reverted that.  The tests are to make sure the component is open or not and I thought we should add the state to the heading, but the heading won't render anyway because the modal is closed and not in the DOM 😵‍💫 😆

File: packages/components/pie-modal/src/index.ts
Comment (ashleynolan) [inline packages/components/pie-modal/src/index.ts:23]: I'm guessing we have some generic tests around the `validPropertyValues` decorator, but is it worth adding specific tests for modal where we pass in the heading types and explicitly test these usecases?

Just to cover if we ever hit an edge case that isn't covered by the generic function tests.

**Review by ashleynolan — COMMENTED** [review]
Looks good – mostly tiny NIT things from me. Otherwise, looks good to merge

File: packages/components/pie-modal/src/index.ts
Comment (raoufswe) [reply packages/components/pie-modal/src/index.ts:20]: i shared about that [here](https://github.com/justeattakeaway/pie/pull/530#discussion_r1224738121)

**Review by fernandofranca — DISMISSED** [review]
Good stuff!!! 🚀

File: packages/components/pie-modal/src/index.ts
Comment (ashleynolan) [reply packages/components/pie-modal/src/index.ts:20]: Thanks – makes sense now 👍

**Review by ashleynolan — APPROVED** [review]

**Review by fernandofranca — APPROVED** [review]

**Pull request merged**
